apps/fipsinstall.c

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/provider.h>
+#include <openssl/params.h>
+#include <openssl/fips_names.h>
+#include "apps.h"
+#include "progs.h"
+
+#define BUFSIZE 4096
+#define DEFAULT_MAC_NAME "HMAC"
+#define DEFAULT_FIPS_SECTION "fips_check_section"
+
+/* Configuration file values */
+#define VERSION_KEY  "version"
+#define VERSION_VAL  "1"
+#define INSTALL_STATUS_VAL "INSTALL_SELF_TEST_KATS_RUN"
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_IN,
+    OPT_OUT,
+    OPT_MODULE,
+    OPT_PROV_NAME,
+    OPT_SECTION_NAME,
+    OPT_MAC_NAME,
+    OPT_MACOPT,
+    OPT_VERIFY
+} OPTION_CHOICE;
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    {OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -provider_name fips"
+     "-section_name fipsinstall -out fips.conf -module ./fips.so"
+     "-mac_name HMAC -macopt digest:SHA256 -macopt hexkey:00"},
+    {"verify", OPT_VERIFY, '-', "Verification mode, i.e verify a config file "
+     "instead of generating one"},
+    {"in", OPT_IN, '<', "Input config file, used when verifying"},
+    {"out", OPT_OUT, '>', "Output config file, used when generating"},
+    {"module", OPT_MODULE, '<', "File name of the provider module"},
+    {"provider_name", OPT_PROV_NAME, 's', "FIPS provider name"},
+    {"section_name", OPT_SECTION_NAME, 's',
+     "FIPS Provider config section name (optional)"},
+    {"mac_name", OPT_MAC_NAME, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm parameters in n:v form. "
+                                "See 'PARAMETER NAMES' in the EVP_MAC_ docs"},
+    {NULL}
+};
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+    size_t outsz = *out_len;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > outsz)
+        goto end;
+    for (;;) {
```

Comment:
```while ((i = BIO_read(...)) != 0)``` seems more clear.  then get rid of 74/75

Reply:
I agree with this

(but in comparison, I'm even more surprised that a simple `(*vtmp++ = 0)` was considered hard to read)

Reply:
The construct goes back to K&R.  In my limited experience, I've never seen a construct like ```|| (*vtmp++ = 0)``` ever.  Can anyone even find another instance of it in openssl?

Reply:
Have you never seen a construct like `*foo = value`? Do you not know that every assignment is an expression with a value, the rvalue? Considering your proposed change, you do know the latter, and we do have the former in our source.

Reply:
Yes I understand all of the above.  Combining it with short-circuiting boolean moves it from standard C to tricky C.  It would be hard to search the source to find it, but I doubt you can find another example *where the code returns zero but is ignored* in a cascading short-circuit operator.  I have never seen that before the other PR that had it.

At a minimum, use the comma operator fer cryin out loud.